src/s3_plugin.js

```diff
@@ -312,4 +322,92 @@ module.exports = class S3Plugin {
       }
     })
   }
+
+  setIndex() {
```

Comment:
I refactored this function and splitted it into two seperate ones, I think it's a lot clearer now.